### DISCUSSION TITLE
ESP32 Audiokit SD to Audiokit Sound Effect Board/pad using setFileFilter()

### DISCUSSION THREAD

**Discussion by vacationvengarl** [pschatzmann/arduino-audio-tools#275] [Q&A] [answered]
Hello again, I'm still working with the ESP32 Audiokit SD-audiokit example. I'm confused about how the code is detecting mp3 files on the SD. I have all my mp3 files in a folder called Music, so I have my SetFilePath= "/Music/"
For me, it randomly selects a file, from my music folder: Test1.mp3, Test2.mp3, and Test3.mp3 . How do I tell how the stream is going  to be ordered? 
I'm also trying to set up conditions to where it only plays 1 mp3 file when condition is met using setFileFilter() but it's not working within void loop(), only works in void setup(). The conditions I'm trying to create is that it plays only 1 mp3 file when a specific button is pressed, like a sound board/pad. I hope I can use setFileFilter() within void loop() because then that function can constantly check to see if the conditions have changed (ie. a different button is pressed so the filter should change to only include the new mp3 file.). 
How do I alter my code to properly implement this sound board/pad? Thank you for your time and consideration btw. 
I still have yet to account for resetting the stream once the mp3 file's audio ends. Below is the void setup/loop sections of my sketch:

void setup() {
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Info);

  //pinMode (PushButton1, INPUT);
  pinMode (PushButton2, INPUT); 
  pinMode (PushButton3, INPUT);
  
  // setup output
  auto cfg = kit.defaultConfig(TX_MODE);
  kit.begin(cfg);

 // setup additional buttons 
  kit.addAction(PIN_KEY4, next);
  kit.addAction(PIN_KEY3, previous);

  // setup player
  player.setVolume(0.7);
  //source.setFileFilter("*Test2*");
  player.begin();
}

void loop() {
//  int button_state1 = digitalRead(PushButton1);
  int button_state2 = digitalRead(PushButton2); 
  int button_state3 = digitalRead(PushButton3);
 /* if(button_state1 == LOW) {
      Active = true;
      source.setFileFilter("*Test1*");
  } */
  if(button_state2 == LOW) {
      Active1 = true;
      source.setFileFilter("*Test2*");
      song = "Test2";
  } 
  if(button_state3 == LOW) {
      Active1 = true;
      source.setFileFilter("*Test3*");
      song = "*Test3*";
  } 
  if(Active1)
  {
      player.copy();
  }
      kit.processActions();
}

**Comment by pschatzmann** [comment, marked as answer]
If you need some ordered files you can implement your own AudioSource subclass or use the AudioSourceCallback class and provide the necessary callbacks. This way you can also handle your conditions in a better way...
